### DISCUSSION TITLE
render mustache template file

### DISCUSSION THREAD

**Discussion by schmurfy** [pulumi/pulumi#8253] [Pulumi in Practice] [answered]
Since I am using typescript I wanted to try to use mustache as a template library to render files but I failed and after multiple attempt I can't find how to make this works, can I do that ?

```ts
let net = new hcloud.Network("private", {
  ipRange: "X.X.X.X/16",
})

let initScript = mustache.render(initScript, {
  hcloud_network: net.id
})

const server = new hcloud.Server(`${name}-server`, {
  // [...]
  userData: cloudInitConfig.rendered
})
```

My issue is that what ends up in `hcloud_network` in the template file is:
```
Calling [toString] on an [Output<T>] is not supported.
    To get the value of an Output<T> as an Output<string> consider either:
    1: o.apply(v => `prefix${v}suffix`)
    2: pulumi.interpolate `prefix${v}suffix`
    See https://pulumi.io/help/outputs for more details.
    This function may throw in a future version of @pulumi/pulumi.
```

So I tried:

```ts
let initScript = mustache.render(initScript, {
  hcloud_network: net.id.apply(s => s)
})
```

With the same result, I also tried to declare a local variable holding the result of apply and pass it to `mustache.render`, same result...
From what I understand there is some kind of promise magic behind the scene with `Input<XX>` and `Output<XX>` but that did not help find the solution.

How can I do that ? is it possible or do I need to stick to the terraform way and use `@pulumi/terraform-template` ?

**Comment by schmurfy** [comment]
I am trying to like pulumi but honestly this is really hard, it is hard to find useful examples of interacting with the outside (anything non terraform/pulumi related), since I am using a real programming language and don't see any point to force me using terraform-template but I would like to at least know if this is a requirement.

**Comment by EvanBoyle** [comment, marked as answer]
@schmurfy you'll need to run `mustace.render` inside of the apply. Something along the lines of:

```ts
let initScript = net.id.apply(s => {
 return mustace.render(initScript, { hcloud_network: s});
});
```

**Reply by schmurfy** [reply]
thanks, I am gonna try that :)
